File: easy/typescript/20-valid-parentheses.ts
function isValid(s: string): boolean {
  const r: Object = {
    "}": "{",
    "]": "[",
    ")": "(",
  };

  const stack: Array<string> = [];

  for (let i = 0; i < s.length; i++) {
    if (r[s[i]]) {
      if (r[s[i]] === stack[stack.length - 1]) {
        stack.pop();
      } else {
        return false;
      }
    } else {
      stack.push(s[i]);
    }
  }

  return stack.length === 0;
}

File: easy/python/20-valid-parentheses.py
class Solution:
    def isValid(s: str) -> bool:
        PARENS_MAP: dict[str, str] = {
            ")": "(",
            "}": "{",
            "]": "[",
        }

        stack: list[str] = []

        for char in s:
            if char in PARENS_MAP:
                if not stack or stack.pop() != PARENS_MAP[char]:
                    return False
            else:
                stack.append(char)

        return not stack
